hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/processor/AsyncTransformUsingContextUnorderedP.java

```diff
@@ -0,0 +1,374 @@
+    private boolean tryProcessSucceeded;
+
+    private Long lastReceivedWm = Long.MIN_VALUE;
+    private long lastEmittedWm = Long.MIN_VALUE;
+    private long minRestoredWm = Long.MAX_VALUE;
+    private int maxAsyncOps;
+    private int asyncOpsCounter;
+
+    /** Temporary collection for restored objects during snapshot restore. */
+    private ArrayDeque<T> restoredObjects = new ArrayDeque<>();
+
+    /**
+     * Constructs a processor with the given mapping function.
+     */
+    private AsyncTransformUsingContextUnorderedP(
+            @Nonnull ContextFactory<C> contextFactory,
+            @Nonnull DistributedBiFunction<? super C, ? super T, CompletableFuture<Traverser<R>>> callAsyncFn,
+            @Nullable C contextObject,
+            @Nonnull Function<? super T, ? extends K> extractKeyFn
+    ) {
+        assert contextObject == null ^ contextFactory.isSharedLocally()
+                : "if contextObject is shared, it must be non-null, or vice versa";
+
+        this.contextFactory = contextFactory;
+        this.callAsyncFn = callAsyncFn;
+        this.contextObject = contextObject;
+        this.extractKeyFn = extractKeyFn;
+    }
+
+    @Override
+    protected void init(@Nonnull Context context) {
+        if (!contextFactory.isSharedLocally()) {
+            assert contextObject == null : "contextObject is not null: " + contextObject;
+            contextObject = contextFactory.createFn().apply(context.jetInstance());
+        }
+        maxAsyncOps = Math.max(1, contextFactory.getMaxPendingCallsPerMember() / context.localParallelism());
+        resultQueue = new ManyToOneConcurrentArrayQueue<>(maxAsyncOps);
+    }
+
+    @Override
+    public boolean isCooperative() {
+        return true;
+    }
+
+    @Override
+    protected boolean tryProcess(int ordinal, @Nonnull Object item) {
+        if (getOutbox().hasUnfinishedItem() && !emitFromTraverser(currentTraverser)) {
+            return false;
+        }
+        @SuppressWarnings("unchecked")
+        T castedItem = (T) item;
+        if (!processItem(castedItem)) {
+            // if queue is full, try to emit and apply backpressure
+            tryFlushQueue();
+            return false;
+        }
+        return true;
+    }
+
+    @CheckReturnValue
+    private boolean processItem(@Nonnull T item) {
+        if (asyncOpsCounter == maxAsyncOps) {
+            return false;
+        }
+        CompletableFuture<Traverser<R>> future = callAsyncFn.apply(contextObject, item);
+        if (future == null) {
+            return true;
+        }
+        asyncOpsCounter++;
+        watermarkCounts.merge(lastReceivedWm, 1L, Long::sum);
+        Long lastWatermarkAtReceiveTime = lastReceivedWm;
+        future.whenComplete(withTryCatch(getLogger(), (r, e) -> {
+            boolean result = resultQueue.offer(tuple3(item, lastWatermarkAtReceiveTime, r != null ? r : e));
+            assert result : "resultQueue doesn't have sufficient capacity";
```

Comment:
Is this intentionally `offer` followed by `assert`, instead of just `add`? A failing offer seems to lose data here.

Reply:
No, I forgot about the `add` method